include/ozo/connection_pool.h

```diff
@@ -8,59 +8,88 @@
 namespace ozo {
 namespace impl {
 
-template <typename ...Ts>
-using connection_pool = yamail::resource_pool::async::pool<connection<Ts...>>;
+template <typename Provider>
+struct get_connection_pool {
+    using type = yamail::resource_pool::async::pool<
+        std::decay_t<decltype(unwrap_connection(std::declval<connectable_type<Provider>&>()))>
+    >;
+};
+
+template <typename Provider>
+using connection_pool = typename get_connection_pool<Provider>::type;
 
-template <typename ...Ts>
+template <typename Provider>
 struct pooled_connection {
-    using handle_type = typename connection_pool<Ts...>::handle;
+    using handle_type = typename connection_pool<Provider>::handle;
+    using underlying_type = typename handle_type::value_type;
 
     handle_type handle_;
 
     pooled_connection(handle_type&& handle) : handle_(std::move(handle)) {}
 
-    decltype(auto) empty() const {return handle_.empty();}
+    bool empty() const {return handle_.empty();}
 
-    void reset(typename handle_type::value_type&& v) {
-        reset(std::move(v));
+    void reset(underlying_type&& v) {
+        handle_.reset(std::move(v));
     }
 
     ~pooled_connection() {
-        if (connection_bad(*this)) {
+        if (!empty() && connection_bad(*this)) {
             handle_.waste();
         }
     }
 
     friend auto& unwrap_connection(pooled_connection& conn) noexcept {
-        return *(conn.handle_);
+        return unwrap_connection(*(conn.handle_));
```

Comment:
You shouldn't do this function noexcept when any of called from there could throw exception, if you don't want to ge unexpected std::terminate. So noexcept specifier should depend on it. Something like this:
```c++
     friend auto& unwrap_connection(pooled_connection& conn) noexcept(noexcept(unwrap_connection(*(conn.handle_)))) {
        return unwrap_connection(*(conn.handle_));
     }
```

Probably we should check our code and fix it in other places. Or do we really need noexcept here?

Reply:
Yes, you right, we need to decide is 'noexcept' needed with unwrap_connection(). Now noexcept presents in connection.h for default unwrap_connection() implementation so that's why it presents here. 

let's fix it in separate PR, I'll make a task.